examples/small-coalton-programs/src/brainfold.lisp

```diff
@@ -0,0 +1,356 @@
+ ;;
+
+ (define print-buffer (cell:new ""))
+  
+ ;;
+ ;; Clean Environment
+ ;;
+  
+ (define (reset-bf)
+   "Resets the bf-environment- pointer, print-buffer, and array."
+   (progn (cell:write! pointer 0)
+	  (cell:write! print-buffer "")
+	  (reset-bf-array)))
+ )
+
+;;;
+;;; Commands (Functions called by Brainfold)
+;;;
+
+(coalton-toplevel
+
+ ;;
+ ;; Navigating through bf-cells (> <)
+ ;;
+  
+ (define (move-right)
+   "Moves the pointer one bf-cell to the right."
+   (cell:increment! pointer))
+
+ (define (move-left)
+   "Moves the pointer one bf-cell to the left."
+   (cell:decrement! pointer))
+
+ ;;
+ ;; Changing bf-cell values (+ -)
+ ;;
+
+ (define (incr)
+   "Increments the value for the current bf-cell."
+   (vec:set! (cell:read pointer)
+	     (+ 1 (current-bf-cell-value))
+	     bf-array)
+   (current-bf-cell-value))
+
+ (define (decr)
+   "Decrements the value for the current bf-cell."
+   (vec:set! (cell:read pointer)
+	     (- (current-bf-cell-value) 1)
+	     bf-array)
+   (current-bf-cell-value))
+
+ ;;
+ ;; Printing Cells (.)
+ ;;
+  
+ (declare print (Ufix -> String))
+ (define (print ascii-value)
+   "Converts a Ufix into its ascii value."
+   (into (make-list (unwrap (char:code-char ascii-value)))))
+
+  
+ (define (print-current-cell)
+   "Prints the current bf-cell"
+   (cell:write! print-buffer (str:concat (cell:read print-buffer) (print (unwrap (tryinto (current-bf-cell-value))))))
+   (current-bf-cell-value))
+
+ ;;
+ ;; Taking Input (,)
+ ;; 
+
+ ;; Currently takes individual characters one at a time as prompted
+
+ (define (character-prompt)
+   "A prompt for obtaining one character as input."
+   (Lisp Char ()
+	 (cl:format cl:*query-io* "Input a character: ")
+	 (cl:force-output cl:*query-io*)
+	 (cl:coerce (cl:read-line cl:*query-io*)
+		    'cl:character)))
+  
+ (define (take-input)
+   "Stores the character at the input-index as an ascii code at the pointer."
+   (vec:set! (cell:read pointer)
+	     (into (char:char-code (character-prompt)))
+	     bf-array)
+   (current-bf-cell-value))
+ )
+
+;;;
+;;; Lexing
+;;; 
+
+(coalton-toplevel
+
+
+ (declare collect-cmds (String -> (List String)))
+ (define (collect-cmds s)
+   "Takes a bf-input string, collects a list of individual command
+characters"
+   (if
+    (== (str:length s) 0)
+    nil
+    (if (list:member (substring s 0 1)
+		     (make-list ">" "<" "+" "-" "." "," "[" "]"))
+	(cons (substring s 0 1)
+	      (collect-cmds (substring s 1 (str:length s))))
+	(collect-cmds (substring s 1 (str:length s))))))
+
+ ;;
+ ;; Defining the Brainfold command type- either a CMD or a Loop
+ ;; containing commands
+ ;;
+
+ (define-type (BF-CMD :t)
+     (CMD String)
+   (BF-LOOP (List (BF-CMD :t))))
+
+ ;;
+ ;; Facilitating returning to commands after the current loop
+ ;;
+  
+ (define (until-end-loop-count bf-commands depth cmd-count)
+   "Cycles through a list of bf-commands and returns the index of the
+character after the corresponding ']'"
+   (cond
```

Comment:
I think this code could be written as a match

Reply:
Match can't match strings right? the list bf-commands is (List String), so I think I would end up with a cond embedded in the match anyways.